crates/blockchain-tree/src/block_indices.rs

```diff
@@ -338,13 +338,58 @@ impl BlockIndices {
     }
 
     /// Returns the block hash of the canonical block with the given number.
-    pub fn canonical_hash(&self, block_number: BlockNumber) -> Option<BlockHash> {
-        self.canonical_chain.get(&block_number).cloned()
+    #[inline]
+    pub fn canonical_hash(&self, block_number: &BlockNumber) -> Option<BlockHash> {
+        self.canonical_chain.canonical_hash(block_number)
     }
 
     /// Returns the block number of the canonical block with the given hash.
+    #[inline]
     pub fn canonical_number(&self, block_hash: BlockHash) -> Option<BlockNumber> {
-        self.canonical_chain.iter().find_map(
+        self.canonical_chain.canonical_number(block_hash)
+    }
+
+    /// get canonical tip
+    #[inline]
+    pub fn canonical_tip(&self) -> BlockNumHash {
+        self.canonical_chain.tip()
+    }
+
+    /// Canonical chain needed for execution of EVM. It should contains last 256 block hashes.
+    #[inline]
+    pub(crate) fn canonical_chain(&self) -> &CanonicalChain {
+        &self.canonical_chain
+    }
+}
+
+/// This keeps track of all blocks of the canonical chain.
+#[derive(Debug, Clone, Default)]
+pub(crate) struct CanonicalChain {
```

Comment:
nit: maybe can go in a separate file?

Reply:
I like this here, because can keep everything private, because only written to by BlockIndices

Reply:
but I think this will grow perhaps, if we add full blocks, so good point, will do

Reply:
It is not a big of deal, it can stay as it is, and if it grows we can always move it.